[Beck] Step3 - Observer Pattern

SangHwi-Back opened the pull request:
## 코멘트 후 수정사항

* 델리게이트 메소드 규칙을 따르지 않는 RectangleViewTapDelegate 프로토콜의 ```changeCurrentSelect(at index: Int?)``` 를 수정.(Naming+)
* NotificationCenter.default.post() 에서 사용하는 userInfo의 키 부분을 하드코딩에서 타입으로 변경.
* Model 객체에 함수/변수명들 중 의미가 광범위하거나 실제 동작과 상이한 이름이 몇 존재하여 변경 필요.(Naming)
* MainScreenAction 타입은 Notification.Name 등으로도 구별 가능하여 삭제.
* Model 중 색상 객체에 사용하려고 만든 propertyWrapper에 init() 추가하여 더 명확한 코드로 개선.(OOP)
* Notification.Name 인 static 변수들은 이제 관련 클래스들이 각자 갖게 됩니다. (Observer Pattern)
* Notification 의 object는 정확히 명시. (Observer Pattern)
* ViewController 에서 액션 별로 옵저버를 등록.(Observer Pattern)
* ViewController 에서 터치한 뷰를 찾고, 프로퍼티를 변경하는 작업도 모델 기준으로 처리.(MVC)
* 투명도 조절 슬라이더에 의해 버튼이 바뀌는 흐름도 모델을 변경한 후 바뀔 수 있도록 소스 변경.(MVC)

## 질문이 있습니다! 🖐️ 

델리게이트 메소드 규칙에 대해 다시 설명 부탁드려도 될까요?   
수업 메모와 PDF를 살펴보았는데 정확한 정보를 찾을 수 없어서 제가 놓친 것 같습니다.   
구글링을 해 보았을 때는 함수명에서 역할을 정확히 명시해야 한다는 정보는 찾을 수 있었는데 그것 말고도 다른 것이 있을까요?   
다음부턴 수업에 더 집중 하도록 하겠습니다....🙇

이전 PR 코멘트입니다.
---

## 작업 내용

* MVC Pattern에서 델리게이트 패턴으로만 요소들을 통신하였던 방식에서 옵저버 방식을 통해 각 요소들의 결합을 느슨하게 만들었습니다.
* NotificationCenter의 여러 사용법에 대해 실습해 보았습니다.

## 고민해 볼 사항

MainScreenViewController의 경우 Plane 모델의 NestedType에 따라 다르게 동작을 하고 있습니다. MainScreenViewController는 Plane에 따라 움직이기 때문에 일정 부분 종속적일 수 밖에 없기는 하지만, 이번 과제의 목표가 ```요소 간 결합을 느슨하게 만들기```라고 생각하고 있는데 작업을 잘 한 것인지 정확히 판단하기 어려웠습니다.

또, NotificationCenter를 사용하고 나니 오히려 viewDidLoad 쪽 코드를 리딩하기가 어려워지는 느낌이 있었습니다. 타입 확장 등을 사용해서 뭔가 탈출구를 찾아보았지만 잘 답이 나오지는 않았습니다. 이 부분은 좀 더 고민해보면서 눈에 잘 들어오는 코드를 작성할지 생각해 봐야 겠습니다.

이번 과제에서는 Notification.Name 갯수를 2개로 만들어서 진행하였습니다. [옵저버_패턴_Wikipedia](https://ko.wikipedia.org/wiki/옵서버_패턴)를 보니 ```이벤트 연쇄```를 주의하라고 나와 있는데, 제 코드가 그러한 것처럼 Notification.Name으로만 post와 observer를 매핑시키면 위험하지 않을까 걱정되기도 하였습니다.

## Keyword

```NotificationCenter``` ```Observer Pattern```

SangHwi-Back committed:
Edit RectRGBColor using propertyWrapper

프로퍼티 래퍼 문법을 RectRGBColor에 적용하였습니다.

SangHwi-Back committed:
Edit Plane separate Rectangle

Plane객체를 Rectangle로부터 완전히 분리하기 위해 MainScreenViewController, ViewController에서
각각 현재 선택된 뷰에 대한 정보를 공유하도록 하였습니다.

SangHwi-Back committed:
Practice Step 3 Observer Pattern

옵저버 패턴을 이용하여 뷰, 뷰컨트롤러, 모델 간의 결합을 느슨하게 만들었습니다.

SangHwi-Back committed:
Resolve Step3

스텝3 마무리를 위해 코드의 응집력을 확인하고 일부 코드를 뷰 컨트롤러와 분리했습니다.
필요한 부분에 주석 등을 달고 README에 완료일시를 기입하였습니다.

SangHwi-Back committed:
Update Access Control

접근 제한자를 추가하지 않은 부분 발견하여 추가하였습니다.
이를 통해 한층 응집력을 올려보고자 하였습니다.

SangHwi-Back committed:
Update README Step3

스텝3의 결과와 추가 학습사항 중 공부한 부분을 업데이트 하였습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
     ///
     /// parentViewController에 전달을 위해 PlaneAdmitDelegate 를 호출하게 됩니다.
-    func changeCurrentSelected(_ rectangle: Rectangle?)
+    func changeCurrentSelected(at index: Int?)
```
(이전에도 썼던가 가물가물하지만) 델리게이트 메소드 규칙을 따르지 않는 것 같아요. 수업 시간에 언급한 적이 있는데 한 번 찾아보세요.
RectangleViewTapDelegate 라면 RectangleView가 보내는 이벤트니까 이름에도, 매개변수에도 규칙이 있어야 합니다

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+    // Plane 객체는 같은 방식으로 Notification을 보내는 것이 좋다고 생각해서
+    // Notification post 만 하는 메소드를 따로 빼서 사용하도록 하였습니다.
+    private func sendNotification(_ model: RectangleProperty, at index: Int, as type: MainScreenAction) {
+        NotificationCenter.default.post(name: .MainScreenAction, object: model, userInfo: ["index": index, "type": type])
```
"index" 나 "type" 문자열도 하드 코딩하기 보다는 상수로 선언해보세요
그래야 post하는 값과 받아서 꺼내는 경우에 같은 키로 원하는 것을 꺼낼 수 있게 됩니다

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+    // MARK: - MainScreenDelegate implementation
+    func addRectangle() {
+        guard
+            let factoryProperty = screenDelegate?.getScreenViewProperty(),
```
factoryProperty는 무엇을 의미하는 지 모르겠습니다. 너무 큰 이름이랄까, 좀 더 구체적인 역할을 표현하면 좋을 것 같네요

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+        else {
+                return
+        }
+        properties.append(property)
```
property 라는 단어도 어떤 속성을 의미하는 것인지 명확하지 않은 것 같아요. 타입이름으로는 좋은데, 변수 이름으로는 좀 더 구체적인 선언이면 좋을 것 같습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+        sendNotification(property, at: properties.endIndex-1, as: .AddButtonPushed)
+    }
+    
+    func setRandomColor(at index: Int) -> RectRGBColor? {
```
이 부분은 좀 어색하네요. `setRandomColor(at index:)`는 어딘가에 랜덤컬러를 넣는건가요? 
set-인데 return을 해서 건네주니까 마치 get- 같기도 하고. 이 함수의 의도를 잘 모르겠어요

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+    }
     
+    // MARK: - RectangleViewTapDelegate implementation
+    func changeCurrentSelected(at index: Int?) {
+        
+        guard let index = index else {
+            NotificationCenter.default.post(name: .MainScreenTouched, object: nil, userInfo: nil)
+            return
+        }
+        let model = properties.count-1 >= index ? properties[index] : nil
+        
+        NotificationCenter.default.post(name: .MainScreenTouched, object: model, userInfo: ["index": index as Any])
```
이 메소드도 좀 낯선 표현입니다. changeCurrentSelected()는 현재 선택을 바꾸는 동작...인거 같은데 그냥 `select(at:)` 이어도 되는걸까요? 조금 과도한 표현같아요.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -12,20 +12,70 @@ protocol RectangleViewTapDelegate {
+    }
+    
+    // MARK: - Plane no using Interface
     func addProperties(_ model: RectangleProperty) {
```
addProperties()는 여러 속성들을 한꺼번에 바꾸는 느낌입니다. Model이 RectangleProperty 인게 마치 사각형 자체를 표현하는 게 아니라 속성들만 모아놓은 느낌인걸까요?

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -55,67 +105,15 @@ final class Plane: RectangleViewTapDelegate {
+    /// root view의 액션 타입
+    ///
+    /// ViewController -> Plane -> MainScreenViewController
+    enum MainScreenAction {
+        /// 무작위 색상 버튼을 선택
+        case ColorButtonPushed
+        /// 투명도 슬라이더 움직임
+        case SliderMoved
+        /// 사각형 추가 버튼을 선택
+        case AddButtonPushed
```
어.. 이런 액션이 있을 수는 있는데 이게 왜 Plane에 있는 걸까요?
Plane에서는 뷰 컨트롤러가 처리할 액션을 직접 갖고 있지 않고, 내부 속성과 관련있는 타입들만 있으면 될 꺼 같은데 이건 입력 뷰에서 처리할 이벤트 액션같은데요?

godrm commented on DrawingApp/DrawingApp/Models/RandomPropertySubTypes.swift:
```diff
@@ -42,50 +43,46 @@ struct RectRGBColor {
     }
+}
+
+@propertyWrapper
+struct ColorValue {
+    private var _value: Double = 0
     
```
어우 프로퍼티 래퍼를 쓰셨네요. 이 정도 노력이면 init() 매개변수 타입을 바꾸는 것도 해보세요 ㅎㅎ

godrm commented on DrawingApp/DrawingApp/TypeExtensions/Notification+Extension.swift:
```diff
@@ -0,0 +1,41 @@
+import UIKit
+
+// MARK: - Predifined Notification Names
+extension Notification.Name {
+    /// Notification name MainScreen(Rectangle etc...) touched
+    static let MainScreenTouched = Notification.Name(rawValue: "MainScreenTouched")
+    /// Notification name MainScreenControl(AddButton, AlphaSlider, ColorButton etc...) touched
+    static let MainScreenAction = Notification.Name(rawValue: "MainScreenAction")
```
이렇게 사용하는 노티는 어느 타입과 관련이 높을까요? 응집도를 높이기 위해서 관련있는 타입에 선언해보세요

godrm commented on DrawingApp/DrawingApp/TypeExtensions/Notification+Extension.swift:
```diff
@@ -0,0 +1,41 @@
+}
+
+// MARK: - Add Observer Each ViewControllers
+extension MainScreenViewController {
+    // handler를 호출하는 쪽에서 정의하도록 한 이유는 아래의 함수가 호출하는 쪽의 응집력을 뺏어간다고 판단했기 때문입니다.
+    func observeMainScreenAction(using handler: @escaping (Notification) -> Void) {
+        NotificationCenter.default.addObserver(
+            forName: .MainScreenAction,
+            object: nil,
+            queue: OperationQueue.main,
+            using: handler
+        )
+    }
+}
```
이 메소드는 왜 MainScreenViewController 와 분리해서 extension으로 구현했는지 궁금합니다. 어떤 의도가 있는건가요? 
주석에 써 있는 응집력을 뺏어간다는 표현은 어떤 의미일까요? handler를 호출과 응집도가 어떤 의미인지 잘 모르겠습니다
handler를 구현하는 쪽에서 이 구현을 모르고 있어야 한다는 걸까요?
혹시 MainScreenViewController 내에 있는 함수들 아닌가요? 같은 타입 내에서는 같이 두는 게 응집도가 높은게 아닐까요?

SangHwi-Back replied:
안녕하세요 JK. 코멘트를 읽고 보니 의도한 바를 제대로 표현하지 못한 메소드인 것 같습니다.

저 **observeMainScreenAction**는 ```옵저버 패턴과 깊은 연관이 있는 Notification.default.addObserver() 메소드를 뷰 컨트롤러 소스코드에서 최대한 제외하고 뷰와 관련된 부분만 남겨보고 싶다.```는 의도에서 만든 것입니다. **handler를 호출부에서 구현**하는 것과 응집도를 연관시킨 이유도 ```ViewController의 프로퍼티를 사용하는 소스코드가 길어질 수록 ViewController와 NotificationCenter의 응집도가 높아진다고 생각```했는데, 의도와는 연관이 없는 작업이었던 것 같습니다.

아래의 코멘트를 모두 읽어보았는데 옵저버를 여럿 등록하는 방법으로 ViewController의 응집도를 더 올려보겠습니다. 언제나 사려깊은 코멘트 감사드립니다.

godrm commented on DrawingApp/DrawingApp/TypeExtensions/Notification+Extension.swift:
```diff
@@ -0,0 +1,41 @@
+    func observeMainScreenTouched(using handler: @escaping (Notification) -> Void) {
+        NotificationCenter.default.addObserver(
+            forName: .MainScreenTouched,
+            object: nil,
```
observer에서 object는 sender의 의미라서 매우 중요한 요소입니다. 
핸들러만큼 중요한 게 저걸 항상 nil로 하는게 좋은걸까요?

godrm commented on DrawingApp/DrawingApp/Views/MainScreenViewController.swift:
```diff
@@ -9,48 +9,39 @@ import UIKit
     var delegate: RectangleViewTapDelegate?
     override func viewDidLoad() {
         super.viewDidLoad()
+        observeMainScreenAction(using: { [weak self] noti in
+            guard let self = self else { return }
+            
+            guard
+                let model = noti.object as? RectangleProperty,
+                let index = noti.userInfo?["index"] as? Int,
+                let type = noti.userInfo?["type"] as? Plane.MainScreenAction
+            else { return }
+            
+            switch type {
+            case .AddButtonPushed:
+                self.view.addSubview(Rectangle(model: model, index: index))
+            case .ColorButtonPushed:
+                self.current?.setBackgroundColor(using: model.rgbValue, alpha: model.alpha)
+            case .SliderMoved:
+                self.current?.setValue(alpha: Float(model.alpha))
+            }
+        })
```
옵저버 핸들러를 이렇게 합쳐서 사용하는군요
제 생각에는 옵저버를 여기서 선언한 액션별로 다루는 게 더 좋을 것 같은데요? 
이렇게 합쳐놓고 switch로 분기하는 게 좋은 구조가 아닙니다. 서로 다른 속성을 갖고 있을꺼고 전달하는 타입도 다르고 호출하는 시점도 다 다를텐데 하나로 합쳐야 하는지 모르겠습니다.

godrm commented on DrawingApp/DrawingApp/Views/MainScreenViewController.swift:
```diff
@@ -63,11 +54,13 @@ final class MainScreenViewController: UIViewController, MainScreenDelegate, UIGe
     // MARK: - UIGestureRecognizerDelegate implementation
-    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
         let touchedView = touches.first?.view as? Rectangle
+        current?.isSelected = false
+        current = touchedView
+        current?.isSelected = true
```
터치한 뷰를 찾는 것도 Rectangle 뷰 기준이 아니라 모델 기준으로 처리해보세요
여기도 입력에 해당하는 것인데 plane으로 이벤트가 전달되야 합니다. 그게 안 보이는 것 같네요

godrm commented on DrawingApp/DrawingApp/Views/ViewController.swift:
```diff
@@ -7,51 +7,63 @@
-        plane.setAlpha(value: sender.value)
+        plane.setAlpha(value: sender.value, at: index)
+        
+        let color = buttonSetRandomColor.backgroundColor
         buttonSetRandomColor.backgroundColor =
-            buttonSetRandomColor.backgroundColor?.withAlphaComponent(Double(sender.value)/RectRGBColor.maxAlpha)
+            color?.withAlphaComponent(Double(sender.value)/RectRGBColor.maxAlpha)
```
sliderAdmitAlphaValueChanged() 처럼 입력에 대한 처리 메소드는 뷰 컨트롤러를 거쳐서 모델까지 들어거는 흐름만 다루며 됩니다
아래있는 버튼 컬러를 바꾸는 것은 출력에 해당하기 때문에 분리해보세요.
출력은 모델에서 뷰 컨트롤러를 거쳐서 뷰까지 가는 흐름을 만들어야 합니다. 그걸 나누는 게 중요한 포인트입니다.

godrm reviewed (changes requested):
옵저버를 사용해서 구조를 나눈 부분에서 어색한 부분들을 의견을 남겼습니다.
입력 흐름과 출력 흐름을 구분해서 개선해보셨으면 합니다.

SangHwi-Back committed:
Update README contents step3 comment

PR 이후 코멘트를 정리 하였습니다.
코멘트를 토대로 소스 코드를 추가 수정합니다.

SangHwi-Back committed:
Refactor Step3 naming and protocol

델리게이트 메소드 규칙을 따르지 않는 프로토콜을 수정하였습니다.
sendNotification(_:,at:,as:)의 하드코딩 부분을 상수로 변환합니다.
addRectangle()의 변수이름이 너무 광범위한 의미라서 변환하였습니다.
setRandomColor(at:)는 set/get을 동시에 하고 있어서 둘 다를 의미할 수 있는 reset으로 변경했습니다.

SangHwi-Back committed:
Refactor step3

ViewController 이름을 MainViewController로 바꾸어서 좀 더 의미 명확하게 하였습니다.
MainScreenAction 타입은 모델에 종속적이 아니므로 MainViewController에 종속적으로 만들었습니다.
RectRGBColor의 PropertyWrapper에 init을 넣고 RectRGBColor designated init을 삭제했습니다.

SangHwi-Back committed:
Refactor Step3 NotificationCenter functionality

NotificationCenter 관련 소스코드를 많이 변경하였습니다.
Notification.Name의 경우 관련 클래스의 클래스 변수로 선언하여 응집력을 높입니다.
각 기능별로 NotificationCenter의 옵저버를 등록하였습니다.
post시 Notification 객체의 이름은 출발지 정보를, object는 실제 알림 보낸 정보를 보냅니다.
뷰 컨트롤러는 모델의 입력 역할에서 완전 분리하고, 출력만을 담당하도록 추가 수정합니다.

SangHwi-Back committed:
Refactor Step3 MainScreen model

사각형 혹은 배경화면을 탭 했을 때 선택된 상태는 로직 자체가 간단하고 뷰와 깊이 연관되어
MainScreenViewController에서 관리하였으나 이를 분리하여 모델 객체에서 관리하고
뷰 컨트롤러에는 Notification만 전송하는 것으로 변경하였습니다.

SangHwi-Back commented:
변경요청 주신 부분에 대해 재수정을 진행하였습니다. 재검토를 부탁드립니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,50 +7,55 @@
+    private func sendNotificationToScreen(using name: Notification.Name, sends model: RectangleProperty, at index: Int) {
+        NotificationCenter.default.post(
+            name: name,
+            object: type(of: self),
```
보통은 object는 인스턴스를 의미하고 type(of:)를 해서 클래스를 지정하지는 않습니다. 참고하세요

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,50 +7,55 @@
     private var properties = [RectangleProperty]()
+    private var selectedIndex: Int?
     
     var screenDelegate: MainScreenDelegate?
```
델리게이트는 그게 누구인지 모르고 연결하려는 겁니다. MainScreen에 대한 Delegate가 혹시 그런 의미라면 조금 다른 의미 같습니다.
여기 델리게이트는 Plane이 무언가 알려주거나 물어볼 대상이 되는 plane의 델리게이트인거고 누구인지 정해진 게 아닙니다.

SangHwi-Back replied:
델리게이트에 대해 질문드린 부분에 대한 답변이군요! 감사합니다.
plane이 소유하는 델리게이트는 단순히 _MainScreenViewController의 델리게이트_ 라는 의미보다는 _실제 무언가를 표현하는 screen의 델리게이트_라는 의미를 표현하고 싶었습니다. 하지만 MainViewController 때문에 제 의도가 명확히 보이지 않는 문제가 있었던 것 같습니다.

MainScreenDelegate를 MainSceneDelegate로 바꾸려고 합니다. MainScene은 MainViewController도 될 수 있고 다른 뷰 컨트롤러 클래스도 될 수 있어서 Plane은 누구인지 알 수 없지만, 자신이 보여주어야 할 MainScene의 책임을 갖고 있음을 예상해볼 수 있는 이름이라고 생각이 듭니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,50 +7,55 @@
-            let factoryProperty = screenDelegate?.getScreenViewProperty(),
-            let property = factory.makeRandomView(as: "Subview #\(properties.count)", property: factoryProperty)
+            let randomRectangleProperty = screenDelegate?.getScreenViewProperty(),
+            let rectangleViewProperty = factory.makeRandomView(as: "Subview #\(properties.count)", property: randomRectangleProperty)
```
이 부분이 좀 이상합니다. 
- screenDelegate?.getScreenViewProperty() 동작에서 델리게이트에게 물어봐서 속성까지 가져와야 하는 걸까요? 
- factory.makeRandomView()는 뷰를 만드는 거라면 Plane에서 할 일이 아닌 것 같은데요? 
Plane이 모델이라면 모델 값만 생성하고 관리하면 되고, 그걸 델리게이트한테 알려주면 필요하면 뷰를 생성하면 되는 게 아닐까요

SangHwi-Back replied:
screenDelegate?.getScreenViewProperty() 의 동작에서 가져오는 속성은 팩토리에서 사용할 속성입니다. 랜덤으로 한 화면 안에 생성되어야 하기 때문에 필연적으로 뷰의 최대 크기를 알지 못하면 랜덤 뷰를 생성하지 못하겠다는 생각에 추가하였습니다.

factory.makeRandomView() 는 뷰를 만들지 않습니다. 대신, Rectangle 생성자에 필요한 모델을 생성합니다.  네이밍으로 인한 오해인 것 같아 수정할 예정입니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,50 +7,55 @@
+        properties.append(rectangleViewProperty)
         
-        sendNotification(property, at: properties.endIndex-1, as: .AddButtonPushed)
+        sendNotificationToScreen(using: MainViewController.addButtonPushed, sends: rectangleViewProperty, at: properties.endIndex-1)
```
이 부분 noti를 보내는 부분도 이상합니다. MainViewController을 알고 보내면 안되구요, 옵저버가 누가 했는지 모르는 상태로 특정한 값이 바뀌었다는 것을 알려주기만 하면 됩니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,50 +7,55 @@
         }
         
-        sendNotification(properties[index], at: index, as: .ColorButtonPushed)
+        sendNotificationToScreen(using: MainViewController.colorButtonPushed, sends: properties[index], at: index)
```
여기도 마찬가지로 MainViewController로 보내는 게 아니라 대상이 누군지 모르지만 post를 해야 합니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -60,23 +65,36 @@ final class Plane: RectangleViewTapDelegate {
-        guard let index = index else {
-            NotificationCenter.default.post(name: .MainScreenTouched, object: nil, userInfo: nil)
-            return
+        var planeNoti = Notification(name: MainScreenViewController.rectangleViewTouched, object: self, userInfo: nil)
```
노티 이름은 받는 타입과 관련된 게 아니라 반대로 보내는 타입과 관련된 겁니다. 
내가 바뀐 걸 알려주는 수단일 뿐입니다.

godrm reviewed (changes requested):
Plane에서 Notification 보내는 부분을 살펴보셔야 할 것 같습니다.
보내는 타입과 관련이 있도록 처리하셔야 합니다. 이 부분만 개선하고 넘어가시죠

SangHwi-Back committed:
Refactor Delegate protocol

델리게이트는 역할에 집중하는 것으로 특정 클래스를 대표해서는 안된다.
MainScreenViewController가 이미 있으므로, 이름을 MainScene으로 변경하여 역할에 더욱 집중하도록 합니다.

SangHwi-Back committed:
Correct wrong typed

오탈자를 수정하였습니다.

SangHwi-Back committed:
Refactor Notification code and Architecture

1. Plane.addRectangle() 메소드에서는 뷰를 생성하지 않지만, 네이밍에 의해 뷰를 생성하는 것처럼 보입니다.
    네이밍을 전체 수정합니다.
2. 모든 Notification은 Plane에서 보냅니다. Notification.Name은 전부 Plane의 클래스 프로퍼티가 됩니다.
3. NotificationCenter로 보내지는 모든 post의 object는 MainViewController의 경우 plane 객체이고,
    MainScreenViewController의 경우 자신의 책임인 MainScreen과 관련된 델리게이트 구현체 중 자신이 참조한 객체입니다.

SangHwi-Back commented:
안녕하세요 JK 추가 수정을 진행하였습니다.

1. NotificationCenter로 post시 object를 type(of: Plane.self)에서 다음과 같이 변경하였습니다.
   * MainScreenViewController의 경우 자신과 관련된 delegate를 구현한 객체(Plane)로 지정하였습니다.
   * MainViewController의 경우 자신이 초기화 한 plane변수를 지정하였습니다.
3. Plane의 addRectangle() 부분이 이상하다고 지적해주셔서 코멘트를 유심히 읽어 보았습니다.   
   해당 메소드에서 실행되는 모든 코드는 사각형 뷰의 모델을 생성하는 부분이었습니다.
   네이밍이 뷰를 만드는 것처럼 보여서 네이밍을 수정하였습니다.

더 수정해야 할 부분이 있다면 가감없이 말씀 부탁드립니다. 늦은 시간 답변 주셔서 다시 한번 감사드립니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -9,63 +9,69 @@ import Foundation
         guard
-            let randomRectangleProperty = screenDelegate?.getScreenViewProperty(),
-            let rectangleViewProperty = factory.makeRandomView(as: "Subview #\(properties.count)", property: randomRectangleProperty)
+            let sceneRect = mainSceneDelegate?.getRect(),
+            let rectangleModel = factory.makeRandomRectangleModel(as: "Subview #\(rectangleModels.count)", rectangleRect: sceneRect)
```
이번에는 좀 더 어떤 의도인지 알 것 같습니다. 이런 ainSceneDelegate?.getRect() 값 경우는 매번 물어봐도 거의 같은 값일텐데 그러면 매번 물어보는 것보다는 한 번에 알려줄 수 있지 않을까요? 또는 매개변수로 넘길 수 있지 않을까도 생각해보세요

godrm commented on DrawingApp/DrawingApp/Views/MainScreenViewController.swift:
```diff
@@ -157,6 +155,6 @@ final class MainScreenViewController: UIViewController, MainScreenDelegate, UIGe
     override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
         let touchedView = touches.first?.view as? Rectangle
-        delegate?.mainScreenDidSelect(at: touchedView?.index)
+        delegate?.mainSceneDidSelect(at: touchedView?.index)
```
이렇게 호출할 때도 delegate가 무엇에 대한 델리게이트인지 의도가 드러나면 좋습니다
Plane의 역할을 표현하는 단어라면 더 좋구요.

SangHwi-Back replied:
JK 질문이 있습니다.

말씀해주신 delegate가 무엇에 대한 의도인지 드러나는 것이 좋다라는 것은   
**"화면이 선택되었다"** 라는 의도가 아닌 **"선택 액션이 일어났으니 해당하는 모델 데이터의 변화를 주어라"** 는 의도를 드러내는 것이 좋다는 말씀이신건가요?

godrm replied:
좋은 질문이네요. 제가 말하고 싶었던 것은 MainScreenViewController의 delegate 자체가 어떤 역할을 하는가 입니다.
그냥 델리게이트보다 좀 더 구체적으로 어떤 것을 위임하고 물어보는 것인가 의도가 들어나면 좋겠다는 겁니다. 
뷰 컨트롤러의 델리게이트라면 무엇을 해야하지? 약간 의아한 부분도 있거든요. 그럼 델리게이트가 아니라 이벤트를 전달해줘야 하는 다른 객체일까? 이렇게 생각이 흘러가게 되더군요

SangHwi-Back replied:
델리게이트를 사용하였다는 것, 전체적인 의도와 세부의도를 나타내는 부분까지 고려하는 것까지 생각하려니 단순히 변수 이름을 delegate로 짓거나 출발지 이름을 딴 protocol 이름은 적절치 못하군요....

MainScreenViewController가 참조하고 있는 델리게이트 패턴의 역할이 무엇인지 다시 한번 고민해보도록 하겠습니다. 답변 감사드립니다!

godrm commented on DrawingApp/DrawingApp/Views/MainViewController.swift:
```diff
@@ -87,7 +83,7 @@ final class MainViewController: UIViewController {
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
         if let vc = segue.destination as? MainScreenViewController {
-            plane.screenDelegate = vc
+            plane.mainSceneDelegate = vc
```
아쉬운 부분은 이런 겁니다. NotificationCenter를 두는 이유가 ViewController가 Plane을 소유하고 있으니 Plane은 상위 타입 ViewController를 어떤 식으로든 모르도록 만들고 싶은데, `plane.mainSceneDelegate`에서 결국 ViewController를 참조하게 되니까요. 그 관계를 다 끊지 못한 부분이 아쉽습니다.

SangHwi-Back replied:
결국 이 코드때문에 완전히 관계를 끊지 못했군요... 다음 스텝 진행 시 이 부분도 함께 작업하여 독립적인 객체들로 만들어보도록 하겠습니다.

godrm commented:
이번단계 고생하셨어요 💪🏻

SangHwi-Back commented:
이번단계 추가로 코멘트 주신 부분 작업과 다음 스텝 작업 시작하겠습니다.

감사합니다!
